### PR TITLE
release v2 use aws-sdk-v2

### PR DESCRIPTION
**Pull request by BobDu**

### CODE REVIEW

**Comment by vieux** [comment]
👍 would be a good idea to have this merged.

**Comment by vieux** [comment]
ping @a8m, I see you're pretty active on GH. Are you interested in merging this?

**Comment by BobDu** [comment]
@a8m ping.

With AWS announcing the end of support for AWS SDK for Go v1 by July 31, 2025 (https://aws.amazon.com/blogs/developer/announcing-end-of-support-for-aws-sdk-for-go-v1-on-july-31-2025/), This upgrade seems to be more necessary now.

Could you take some time to review this PR and release a new version?

File: aggregator.go [hunk 1]
```diff
@@ -53,28 +54,28 @@ func (a *Aggregator) Put(data []byte, partitionKey string) {
 // that compatible with the KCL's deaggregation logic.
 //
 // If you interested to know more about it. see: aggregation-format.md
-func (a *Aggregator) Drain() (*k.PutRecordsRequestEntry, error) {
+func (a *Aggregator) Drain() (ktypes.PutRecordsRequestEntry, bool, error) {
 	if a.nbytes == 0 {
```
Comment (a8m) [inline aggregator.go:58]: Why did you change the signature and not just changed the type?

File: .gitignore [hunk 2]
```diff
@@ -1 +1,2 @@
 draft
+.idea
```
Comment (a8m) [inline .gitignore:2]: Use global .gitignore instead. This file should be removed from here completely

File: go.mod [hunk 3]
```diff
@@ -1,12 +1,17 @@
 module github.com/a8m/kinesis-producer
 
+go 1.16
```
Comment (a8m) [inline go.mod:3]: upgrade to 1.23

File: producer.go [hunk 4]
```diff
@@ -215,26 +219,26 @@ func (p *Producer) loop() {
...
 
-func (p *Producer) drainIfNeed() (*kinesis.PutRecordsRequestEntry, bool) {
+func (p *Producer) drainIfNeed() (ktypes.PutRecordsRequestEntry, bool) {
 	p.RLock()
 	needToDrain := p.aggregator.Size() > 0
 	p.RUnlock()
 	if needToDrain {
 		p.Lock()
-		record, err := p.aggregator.Drain()
+		record, has, err := p.aggregator.Drain()
```
Comment (a8m) [inline producer.go:228]: Let's not mix logic change with SDK upgrade. Also, if the previous code uses pointers, keep using pointers.

File: producer_test.go [hunk 5]
```diff
@@ -20,7 +22,7 @@ type clientMock struct {
 	incoming  map[int][]string
 }
 
-func (c *clientMock) PutRecords(input *k.PutRecordsInput) (*k.PutRecordsOutput, error) {
+func (c *clientMock) PutRecords(ctx context.Context, input *k.PutRecordsInput, optFns ...func(*k.Options)) (*k.PutRecordsOutput, error) {
```
Comment (a8m) [inline producer_test.go:25]:
```suggestion
func (c *clientMock) PutRecords(_ context.Context, input *k.PutRecordsInput, _ ...func(*k.Options)) (*k.PutRecordsOutput, error) {
```

File: producer.go [hunk 6]
```diff
@@ -243,8 +247,8 @@ func (p *Producer) flush(records []*kinesis.PutRecordsRequestEntry, reason strin
 
 	for {
 		p.Logger.Info("flushing records", LogValue{"reason", reason}, LogValue{"records", len(records)})
-		out, err := p.Client.PutRecords(&kinesis.PutRecordsInput{
-			StreamName: &p.StreamName,
+		out, err := p.Client.PutRecords(context.TODO(), &kinesis.PutRecordsInput{
```
Comment (a8m) [inline producer.go:250]:
```suggestion
		out, err := p.Client.PutRecords(context.Background(), &kinesis.PutRecordsInput{
```

**Review by a8m — COMMENTED** [review]
Thanks for the contribution, @BobDu.

Sorry for my delayed review. I wrote this package nine years ago when I built my product on top of Kinesis, but I currently do not use this service, and therefore, this package is not actively maintained.

**Comment by BobDu** [comment]
https://github.com/aws/aws-sdk-go/blob/825250a3f2f45ff9322c4a9ae2dd96e5bdb93ea4/service/kinesis/api.go#L8237

https://github.com/aws/aws-sdk-go-v2/blob/65023f312c12ce2707836e552ab55bb4f03f3345/service/kinesis/api_op_PutRecords.go#L92

Due to the modification of the `Records` field type in `PutRecordsInput` by `aws-go-sdk-v2`
`Records []*PutRecordsRequestEntry` => `Records []types.PutRecordsRequestEntry`

So, I have also modified some function signatures to no longer use pointer slices.
However, for non-slice type parameters, I have reverted to the previous behavior.

**Comment by BobDu** [comment]
Additionally, I have updated the version of the gRPC library and other dependencies.

All comment are resolved. @a8m Please review it again.

**Comment by BobDu** [comment]
Additionally, I have added support for the kinesis `StreamARN`

**Comment by a8m** [comment]
> Additionally, I have added support for the kinesis `StreamARN`

Please, let’s not mix changes in the same PR. I want to approve and merge the first one.

**Comment by BobDu** [comment]
> Please, let’s not mix changes in the same PR. I want to approve and merge the first one.

Sorry, I revert this change.
If you think it is necessary, I can revert update protobuf dependencies commit same.

And I will create other PR.  @a8m

**Review by a8m — APPROVED** [review]
💯 

Thanks for the contribution, @BobDu
